Allow code interface to implement a stopping criteria to prematurely stop a simulation if criteria are met

alfoa opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #2468

##### What are the significant changes in functionality due to this change request?
This PR implements a method for code interfaces to stop a simulation
if criteria (i.e. user defined criteria) are met.
The code interfaces can now implement a new optional method  onlineStopCriteria:
If present, it is is called by RAVEN during the simulation.
It is intended to provide means for the code interface to monitor the execution of a run
and stop it if certain criteria are met (defined at the code interface level).
For example, the underlying code interface can check for a figure of merit in the output file
produced by the driven code and stop the simulation if that figure of merit is outside a certain interval
(e.g. Pressure > 2 bar, stop otherwise, continue).
If the simulation is stopped because of this check, the return code is set artificially to 0 (normal termination) and
the ``checkForOutputFailure''  method (if present) is not called. So the simulation is considered to be successful.
This method can be used to inquire an external function (user-implemented) that can be used to implement
a stop logic based on one or multiple figure of merits.

This PR implements this method in the RAVEN mechanics and also provide an example (implementation)
in the Serpent Code Interface. 

In addition: a variable is created that monitors the reason for the code stopping
Options are:
- Normal: Normal termination
- Timeout: Timeout of the simulation in the driven code
- StoppingCondtion: Normal Termination, Stopping condition triggered


For peace of mind, I also created folders for the code interface tests (each / code) except the obsolete ones.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

alfoa commented on ravenframework/Models/Code.py:
```diff
@@ -610,17 +610,43 @@ def evaluateSample(self, myInput, samplerType, kwargs):
-          break
+    if self.maxWallTime is not None or self.code._hasOnlineStopCriteriaCheck:
+      stoppingCriteriaTimeInterval = self.code.getOnlineStopCriteriaTimeInterval() if self.code._hasOnlineStopCriteriaCheck else None
+
+
```
```suggestion
```

alfoa commented on ravenframework/Models/Code.py:
```diff
@@ -610,17 +610,43 @@ def evaluateSample(self, myInput, samplerType, kwargs):
+              process.kill()
+              process.returncode = 0
+            currentTime = time.time()
+
```
```suggestion
```

alfoa commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -336,7 +336,6 @@ def pcmTdep(featData, msrData, targData, recError):
       """
       yExp = np.array(featData)
       yApp = np.array(targData)
-      yExpMsr = np.array(msrData)
```
unused

alfoa commented on ravenframework/utils/TreeStructure.py:
```diff
@@ -17,7 +17,6 @@
 """
 
 import xml.etree.ElementTree as ET
-import os, sys
```
not used

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/2808162/) on 6155f40 : invalidated by @joshua-cogliati-inl

failed in fetch SRAW

alfoa commented on ravenframework/CodeInterfaceBaseClass.py:
```diff
@@ -271,6 +282,35 @@ def checkForOutputFailure(self, output, workingDir):
+      @ Out, continueSim, bool, True if the job needs to continue being executed, False if it needs to be stopped
+    """
+    continueSim = self.onlineStopCriteria(command, output, workingDir)
+    assert isinstance(continueSim, bool), f"First return argument of 'onlineStopCriteria' (for code interface {self.type}) must be a bool, got {type(continueSim)}!"
```
```suggestion
    assert isinstance(continueSim, bool), f"Return argument of 'onlineStopCriteria' (for code interface {self.type}) must be a boolean, got {type(continueSim)}!"
```

alfoa commented on ravenframework/CodeInterfaceClasses/Generic/GenericCodeInterface.py:
```diff
@@ -84,7 +84,6 @@ def generateCommand(self,inputFiles,executable,clargs=None, fargs=None, preExec=
     #if len(fargs['output'])<1 and 'output' not in clargs.keys():
     #  raise IOError('No output file was specified, either in clargs or fileargs!')
     #check all required input files are there
-    inFiles=inputFiles[:]
```
not used

alfoa commented on ravenframework/CodeInterfaceClasses/SERPENT/SerpentInterface.py:
```diff
@@ -16,12 +16,17 @@
 """
 #External Modules--------------------begin
 import os
+from xml.etree import ElementTree
```
```suggestion
```

alfoa commented on ravenframework/CodeInterfaceClasses/SERPENT/SerpentInterface.py:
```diff
@@ -54,6 +58,9 @@ def __init__(self):
     self.nVolumePoints = None
+    # function to be evaluated for the stopping condition if any
+    self.stoppingCriteriaFunction = None
+    self.stoppingCriteriaFunctionNode = None
```
```suggestion
    self._stoppingCriteriaFunctionNode = None
```

alfoa commented on ravenframework/CodeInterfaceClasses/SERPENT/SerpentInterface.py:
```diff
@@ -109,6 +117,8 @@ def _readMoreXML(self,xmlNode):
                         f'{", ".join(op.serpentOutputAvailableTypes)}!!')
       self._fileTypesToRead += serpentFileTypes
+    # we just store it for now because we need the runInfo to be applied before parsing it
+    self.stoppingCriteriaFunctionNode = xmlNode.find("stoppingCriteriaFunction")
```
```suggestion
    self._stoppingCriteriaFunctionNode = xmlNode.find("stoppingCriteriaFunction")
```

alfoa commented on ravenframework/CodeInterfaceClasses/SERPENT/SerpentInterface.py:
```diff
@@ -117,6 +127,24 @@ def initialize(self, runInfo, oriInputFiles):
       @ Out, None
     """
+    super().initialize(runInfo, oriInputFiles)
+    if self.stoppingCriteriaFunctionNode is not None:
```
```suggestion
    if self._stoppingCriteriaFunctionNode is not None:
```

alfoa commented on ravenframework/CodeInterfaceClasses/SERPENT/SerpentInterface.py:
```diff
@@ -117,6 +127,24 @@ def initialize(self, runInfo, oriInputFiles):
+      functs = InputNode('Functions')
+      # change tag name to External (to use the parser)
+      # this is very ugly but works fine
+      self.stoppingCriteriaFunctionNode.tag = 'External'
+      functs.append(self.stoppingCriteriaFunctionNode)
```
```suggestion
      self._stoppingCriteriaFunctionNode.tag = 'External'
      functs.append(self._stoppingCriteriaFunctionNode)
```

moosebuild commented:
Job [Test qsubs sawtooth](https://civet.inl.gov/job/2811315/) on e9d6f17 : invalidated by @alfoa

alfoa commented:
For peace of mind, I also created folders for the code interface tests (each / code) except the obsolete ones.

alfoa commented:
Only new tests are in "tests/framework/CodeInterfaceTests/Generic" (for the new stopping condition) and "tests/framework/CodeInterfaceTests/Serpent" (for the same thing). All the other tests have been only moved to dedicated subfolders.

wangcj05 commented on doc/user_manual/code_interfaces/generic.tex:
```diff
@@ -71,7 +71,7 @@ \subsection{Generic Interface}
       <clargs type='input'   arg='-i' extension='.one'/>
       <clargs type='input'   arg='-a' extension='.two'/>
-      <clargs type='output'  arg='-o'/>
+      <clargs type='output'  arg='-o'/> 
```
end trail whitespace?

alfoa replied:
fixed

wangcj05 commented on doc/user_manual/code_interfaces/generic.tex:
```diff
@@ -156,3 +156,90 @@ \subsection{Generic Interface}
+\\To activate such capability, the following addition XML nodes can be inputted:
+\begin{itemize}
+   \item \xmlNode{onlineStopCriteriaTimeInterval}, \xmlDesc{float, optional parameter}, time frequency (in seconds)) at which  the stopping function (if inputted)
+                                                            is inquired (stoppingCriteriaFunction). \default{10}.
```
remove the extra ")" after seconds

wangcj05 commented on doc/user_manual/code_interfaces/generic.tex:
```diff
@@ -156,3 +156,90 @@ \subsection{Generic Interface}
+                                                            function to halt the simulation if a certain condition is
+                                                            met. The criterion or criteria is or are defined through an external RAVEN python function. The function must return a ``bool'':
+                                                            \begin{itemize}
+                                                              \item True, if the simulation can continue (i.e. the criteria are not met)
+                                                              \item False, if the simulation must STOP (i.e. the criteria are met)
```
I may prefer return "True" when the criteria are met here.

alfoa replied:
perfect fixed.

wangcj05 commented on doc/user_manual/code_interfaces/generic.tex:
```diff
@@ -156,3 +156,90 @@ \subsection{Generic Interface}
+        15.0
+      </onlineStopCriteriaTimeInterval>
+      <!-- use the function named criteria in the python file stoppingCriteria.py --> 
+      <stoppingCriteriaFunction file="stoppingCriteria.py" name="criteria">
+         <!-- 
+           variables below must be retrievable
+           from the generic code interface csv output 
+         -->
+        <variables>var1, var2 </variables>
+      </stoppingCriteriaFunction>
```
This is a more consistent question. Inside RAVEN, we use <Functions> block to define all functions, and then reference the function name in corresponding entities so that we can use it. In this implementation, the whole function is defined within the entity. I may prefer we keep the consistency with current RAVEN function structure. What do you think? The following is the example: 
<img width="681" alt="image" src="https://github.com/user-attachments/assets/697d8d1d-87ec-4f25-8f96-7de75a7e6af9" />

alfoa replied:
@wangcj05 I addressed your comment using the assembler object system. let me know what you think.

wangcj05 reviewed (changes requested):
@alfoa I have design questions for you to consider.

moosebuild commented:
Job [Test qsubs sawtooth](https://civet.inl.gov/job/2820261/) on e3c4b9c : invalidated by @joshua-cogliati-inl

failed in fetch Directory not empty

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/2821166/) on fed160a : invalidated by @joshua-cogliati-inl

failed in fetch Directory not empty

wangcj05 reviewed (changes requested):
@alfoa Thanks a lot for your great efforts to make this useful changes. I have several suggestions for you to consider. In addition, I do have a question about how you perform the testing on your method: I guess we can not really test it for most of our code interfaces since we can not run the code directly. I think we can only test it with generic code interface, since we would like to see the code can stop early, should we make corresponding changes in the generic code interface test for both code execution time and the output results? At least we can see with/without the stopCriteriaFunction, the returned results are different. Please let me know what do you think.

alfoa commented on doc/user_manual/couplingAcode.tex:
```diff
@@ -354,6 +355,70 @@ \subsubsection{Method: \texttt{finalizeCodeOutput}}
+ 
+ It is important to notice that this method can leverage any  external function defined in the  \xmlNode{Functions} block (see Section~\ref{sec:functions}) 
+ as long as it returns a boolean.
+ Within the \xmlNode{Code} XML node, the user can already link such function thorugh the following XML node:
```
fixed

alfoa commented on ravenframework/CodeInterfaceBaseClass.py:
```diff
@@ -271,6 +300,35 @@ def checkForOutputFailure(self, output, workingDir):
+    """
+      Method to get the time interval (frequency) of how often the onlineStopCriteriaCheck needs to be inquired.
+      @ In, None
+      @ Out,
```
fixed

alfoa commented on ravenframework/Models/Code.py:
```diff
@@ -610,17 +636,41 @@ def evaluateSample(self, myInput, samplerType, kwargs):
+        currentTime = time.time()
+        timeout = currentTime + self.maxWallTime
+        while True:
+          time.sleep(0.5)
```
this is in the loop for the "wall time" check.

alfoa replied:
combined loops

alfoa commented on ravenframework/Models/Code.py:
```diff
@@ -610,17 +636,41 @@ def evaluateSample(self, myInput, samplerType, kwargs):
+          process.poll()
+          stopSim = self.code.onlineStopCriteriaCheck(command, codeLogFile, metaData['subDirectory'])
+          if stopSim:
+            self.raiseAMessage(f'Code "{self.code.name}" (type: "{self.code.name}") triggered a stopping criteria to stop the run. Return code is set to 0!')
```
sorry this was supposed to be the warning for the timeout check. Fixing.

alfoa replied:
combined loops

alfoa commented:
> @alfoa Thanks a lot for your great efforts to make this useful changes. I have several suggestions for you to consider. In addition, I do have a question about how you perform the testing on your method: I guess we can not really test it for most of our code interfaces since we can not run the code directly. I think we can only test it with generic code interface, since we would like to see the code can stop early, should we make corresponding changes in the generic code interface test for both code execution time and the output results? At least we can see with/without the stopCriteriaFunction, the returned results are different. Please let me know what do you think.

Hey Congjian. Yeah currently the only "Active" test for this feature is in the GenericCode interface since it can use a "code" to be run (that's why I reworked the "poly" code to use a sleep time to slow down the calculation). I was thinking that if the stopping condition is met, we can set a flag to TRUE and add it in the realization output (that then can be exported as a CSV)?

Let me quickly implement it and if you don't like it we can discard the commit.

alfoa commented:
> > @alfoa Thanks a lot for your great efforts to make this useful changes. I have several suggestions for you to consider. In addition, I do have a question about how you perform the testing on your method: I guess we can not really test it for most of our code interfaces since we can not run the code directly. I think we can only test it with generic code interface, since we would like to see the code can stop early, should we make corresponding changes in the generic code interface test for both code execution time and the output results? At least we can see with/without the stopCriteriaFunction, the returned results are different. Please let me know what do you think.
> 
> Hey Congjian. Yeah currently the only "Active" test for this feature is in the GenericCode interface since it can use a "code" to be run (that's why I reworked the "poly" code to use a sleep time to slow down the calculation). I was thinking that if the stopping condition is met, we can set a flag to TRUE and add it in the realization output (that then can be exported as a CSV)?
> 
> Let me quickly implement it and if you don't like it we can discard the commit.

basically the idea is to add a metadata with the "Termination Reason":
- Normal Termination
- Error
- Timeout
- User-stopping condition

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/2821842/) on d5dbb53 : invalidated by @joshua-cogliati-inl

failed in fetch Directory not empty

alfoa commented:
> > > @alfoa Thanks a lot for your great efforts to make this useful changes. I have several suggestions for you to consider. In addition, I do have a question about how you perform the testing on your method: I guess we can not really test it for most of our code interfaces since we can not run the code directly. I think we can only test it with generic code interface, since we would like to see the code can stop early, should we make corresponding changes in the generic code interface test for both code execution time and the output results? At least we can see with/without the stopCriteriaFunction, the returned results are different. Please let me know what do you think.
> > 
> > 
> > Hey Congjian. Yeah currently the only "Active" test for this feature is in the GenericCode interface since it can use a "code" to be run (that's why I reworked the "poly" code to use a sleep time to slow down the calculation). I was thinking that if the stopping condition is met, we can set a flag to TRUE and add it in the realization output (that then can be exported as a CSV)?
> > Let me quickly implement it and if you don't like it we can discard the commit.
> 
> basically the idea is to add a metadata with the "Termination Reason":
> 
> * Normal Termination
> * Error
> * Timeout
> * User-stopping condition

@wangcj05 I implemented and added the test that checks for it

wangcj05 commented:
[like]  Congjian Wang reacted to your message:
________________________________
From: Andrea Alfonsi - NuCube ***@***.***>
Sent: Wednesday, April 16, 2025 3:18:21 PM
To: idaholab/raven ***@***.***>
Cc: Congjian Wang ***@***.***>; Mention ***@***.***>
Subject: Re: [idaholab/raven] Allow code interface to implement a stopping criteria to prematurely stop a simulation if criteria are met (PR #2470)


@alfoa<https://github.com/alfoa> Thanks a lot for your great efforts to make this useful changes. I have several suggestions for you to consider. In addition, I do have a question about how you perform the testing on your method: I guess we can not really test it for most of our code interfaces since we can not run the code directly. I think we can only test it with generic code interface, since we would like to see the code can stop early, should we make corresponding changes in the generic code interface test for both code execution time and the output results? At least we can see with/without the stopCriteriaFunction, the returned results are different. Please let me know what do you think.

Hey Congjian. Yeah currently the only "Active" test for this feature is in the GenericCode interface since it can use a "code" to be run (that's why I reworked the "poly" code to use a sleep time to slow down the calculation). I was thinking that if the stopping condition is met, we can set a flag to TRUE and add it in the realization output (that then can be exported as a CSV)?
Let me quickly implement it and if you don't like it we can discard the commit.

basically the idea is to add a metadata with the "Termination Reason":

  *   Normal Termination
  *   Error
  *   Timeout
  *   User-stopping condition

@wangcj05<https://github.com/wangcj05> I implemented and added the test that checks for it

—
Reply to this email directly, view it on GitHub<https://github.com/idaholab/raven/pull/2470#issuecomment-2809921097>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABX3L36ODFBTBS4D5SPTOX32ZZYD3AVCNFSM6AAAAAB2ZLLTOSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBZHEZDCMBZG4>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

[https://avatars.githubusercontent.com/u/9202671?s=20&v=4]alfoa left a comment (idaholab/raven#2470)<https://github.com/idaholab/raven/pull/2470#issuecomment-2809921097>

@alfoa<https://github.com/alfoa> Thanks a lot for your great efforts to make this useful changes. I have several suggestions for you to consider. In addition, I do have a question about how you perform the testing on your method: I guess we can not really test it for most of our code interfaces since we can not run the code directly. I think we can only test it with generic code interface, since we would like to see the code can stop early, should we make corresponding changes in the generic code interface test for both code execution time and the output results? At least we can see with/without the stopCriteriaFunction, the returned results are different. Please let me know what do you think.

Hey Congjian. Yeah currently the only "Active" test for this feature is in the GenericCode interface since it can use a "code" to be run (that's why I reworked the "poly" code to use a sleep time to slow down the calculation). I was thinking that if the stopping condition is met, we can set a flag to TRUE and add it in the realization output (that then can be exported as a CSV)?
Let me quickly implement it and if you don't like it we can discard the commit.

basically the idea is to add a metadata with the "Termination Reason":

  *   Normal Termination
  *   Error
  *   Timeout
  *   User-stopping condition

@wangcj05<https://github.com/wangcj05> I implemented and added the test that checks for it

—
Reply to this email directly, view it on GitHub<https://github.com/idaholab/raven/pull/2470#issuecomment-2809921097>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABX3L36ODFBTBS4D5SPTOX32ZZYD3AVCNFSM6AAAAAB2ZLLTOSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBZHEZDCMBZG4>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

wangcj05 reviewed (approved):
@alfoa Thanks for the great contribution and accommodate my requests. The test is exactly what I'm looking for.

wangcj05 commented:
checklist is good. PR can be merged.

alfoa commented:
@wangcj05 thanks a lot!!

wangcj05 commented on ravenframework/CodeInterfaceClasses/Generic/GenericCodeInterface.py:
```diff
@@ -178,3 +188,28 @@ def createNewInput(self,currentInputFiles,origInputFiles,samplerType,**Kwargs):
+      @ Out, stopSim, bool, True if the job needs to stop being executed, False if it needs to continue to be executed
+    """
+    stopSim = False
+    if self.stoppingCriteriaFunction is not None:
+      outCsv = os.path.join(workDir,output+".csv")
+      if os.path.exists(outCsv):
+        df = pd.read_csv(outCsv)
+        results =  dict((header, np.array(df[header])) for header in df.columns)
+        stopSim = self.stoppingCriteriaFunction.evaluate(self.stoppingCriteriaFunction.name,results)
```
@alfoa FYI, we observed a failure on windows for the testing of this capability. The reason is: the output is created, but no data has been written yet, which means df[header] is empty, but the stoppingCriteriaFunction got evaluated with empty data which causes the failure of some jobs. We can either fix it here or update the stoppingCriteriaFunction to check it. Do you have some suggestions?
